routing/road_graph.hpp

```diff
@@ -99,17 +111,74 @@ class IRoadGraph
   };
 
   /// This class is responsible for loading edges in a cross.
-  /// It loades only outgoing edges.
-  class CrossEdgesLoader
+  class ICrossEdgesLoader
   {
   public:
-    CrossEdgesLoader(m2::PointD const & cross, TEdgeVector & outgoingEdges);
+    ICrossEdgesLoader(m2::PointD const & cross, IRoadGraph::Mode mode, TEdgeVector & edges)
+      : m_cross(cross), m_mode(mode), m_edges(edges)
+    {
+    }
+    virtual ~ICrossEdgesLoader() = default;
 
-    void operator()(FeatureID const & featureId, RoadInfo const & roadInfo);
+    void operator()(FeatureID const & featureId, RoadInfo const & roadInfo)
+    {
+      LoadEdges(featureId, roadInfo);
+    }
 
   private:
+    virtual void LoadEdges(FeatureID const & featureId, RoadInfo const & roadInfo) = 0;
+
+  protected:
+    template <typename THeadFn, typename TTailFn>
+    void ForEachEdge(RoadInfo const & roadInfo, THeadFn && headFn, TTailFn && tailFn)
```

Comment:
These two lambdas (headFn and tailFn) contain different code inside. I'm not sure if I add a boolean flag (like below) it makes the code more understandable. Are you talking about something like that?

```c++

    template <typename TAddEdgeFn>
    void ForEachEdge(RoadInfo const & roadInfo, TAddEdgeFn && addEdgeFn)
    {
      for (size_t i = 0; i < roadInfo.m_points.size(); ++i)
      {
        m2::PointD const & p = roadInfo.m_points[i];

        if (!my::AlmostEqualAbs(m_cross, p, kPointsEqualEpsilon))
          continue;

        if (i > 0)
        {
          //               p
          // o------------>o
          TAddEdgeFn(i, p, false);
        }
        if (i < roadInfo.m_points.size() - 1)
        {
          // p
          // o------------>o
          addEdgeFn(i, p, true);
        }
      }
    }

  ForEachEdge(roadInfo, [&featureId, &roadInfo, this](size_t i, m2::PointD const & p, bool head)
  {
    if (head)
    {
      ASSERT_LESS(i, roadInfo.m_points.size() - 1, ());
      m_edges.emplace_back(featureId, true /* forward */, i, p, roadInfo.m_points[i + 1]);
    }
    else
    {
      ASSERT_LESS(i, roadInfo.m_points.size(), ());
      if (roadInfo.m_bidirectional || m_mode == IRoadGraph::Mode::IgnoreOnewayTag)
        m_edges.emplace_back(featureId, false /* forward */, i - 1, p, roadInfo.m_points[i - 1]);
    }
  });

```